### DISCUSSION TITLE
x-on:click not working without an x-data on el or parent el.

### DISCUSSION THREAD

**Discussion by joshmanders** [alpinejs/alpine#3805] [7. Bugs (archived)] [answered]
I have been wracking my brain all day trying to figure out why my app I am building using htmx and the `hx-boost="true"` directive wasn't letting my `x-on:click="toggleDarkMode"` on a button wasn't working. 

`toggleDarkMode` is a vanilla JS function in my `<head />` that just adds/removes a class from `<html />` using `document.querySelector()` so the only thing that Alpine is doing is initalizing and should be binding to that `<button />` but it wasn't. While trying to debug I noticed that if I navigated to another page it magically started working again. But only under htmx' boosted pages, if it was a full page reload then it never worked.

So I set out to get a minimal reproduction and I have one here: https://codepen.io/joshmanders/pen/rNoRZOG?editors=1011

As you can see nothing on the page except two buttons and the script tag. The first button uses `x-on:click` to console log `no-data` and the second has a `<div x-data="{}">` wrapping it and console logging `x-data`. If you open the console in Codepen and observe the output you should see 3 logs after clicking both buttons once. Alpine initialized message and logs for each button, but as you'll see, the first one doesn't work.

Is this a bug or do I not realize that a root `x-data` needs to exist before Alpine starts doing its thing?

**Comment by ekwoka** [comment, marked as answer]
It's in the docs.

x-data defines the scope of an Alpine component.

Anything not inside x-data (except x init as an exception) is not checked as being Alpine.

**Reply by joshmanders** [reply]
Ah well I must have missed that part haha.

Any reason it doesn't just attach to any `x-*` directive? Seems weird to need to do `<button x-data x-on:click="alert('foo')">Click Me</button>`

**Reply by ekwoka** [reply]
Because of how the data is built for the components, and the process of looking through the Dom.

It would be a ton of extra work to force on every app.


You can put `x-data` on the body if you want to opt into that behavior.

**Reply by joshmanders** [reply]
Yeah that's what I ended up doing. Thanks for all your hard work and help. Alpine is great!

**Reply by sibeduli** [reply]
i cannot find it inside the docs section of 'x-on' perhaps it's hidden somewhere @ekwoka ?

**Reply by sibeduli** [reply]
however, in the docs section of `x-data' it says 'Everything in Alpine starts with the x-data directive.' this could be interpreted as many things

**Reply by ekwoka** [reply]
It could use some clarification, for sure.

🤔

Maybe a more robust getting started.

**Reply by DevTakao** [reply]
I think this should be a really big font size and also mention that it is also needed by many plugin directives to work such as x-sort. Because in x-sort examples like [this](https://alpinejs.dev/plugins/sort#basic-usage), there is no x-data in the code so I just assumed it wasn't needed and ended up wasting a lot of time.

**Reply by ekwoka** [reply]
I did make a PR to add a note to every page. 

https://github.com/alpinejs/alpine/pull/4307

It was merged. But I guess i didn't do it on x-sort.

Idk man. People got a read
